core/trino-main/src/main/java/io/trino/sql/planner/optimizations/SkewJoinOptimizer.java

```diff
@@ -0,0 +1,440 @@
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, Map<Symbol, SkewedContext> context)
+        {
+            // TODO : how do range joins behave ?
+            Map<Symbol, SkewedContext> leftCtx = new HashMap<>();
+            Map<Symbol, SkewedContext> rightCtx = new HashMap<>();
+
+            PlanNode leftReWritten = node.getLeft().accept(this, leftCtx);
+            PlanNode rightReWritten = node.getRight().accept(this, rightCtx);
+
+            JoinNode newJoin = this.joinRewriter.rewriteJoin(node, leftReWritten, rightReWritten, Stream.empty());
+
+            // TODO: Currently we only handle the case when the skewed symbol is on the probe side. lets also handle the stream side.
+            if (leftCtx.size() > 0) {
+                Optional<JoinNode.EquiJoinClause> maybeClause = newJoin.getCriteria().stream()
+                        .filter(clause -> leftCtx.containsKey(clause.getLeft()))
+                        .findFirst();
+
+                if (maybeClause.isPresent()) {
+                    JoinNode.EquiJoinClause skewedClause = maybeClause.get();
+                    return this.joinRewriter.rewriteSkewedJoin(newJoin, skewedClause, leftCtx.get(skewedClause.getLeft()));
+                }
+            }
+
+            return newJoin;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitSort(SortNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, Map<Symbol, SkewedContext> context)
+        {
+            return passThrough(node, context);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, Map<Symbol, SkewedContext> context)
+        {
+            /*
+                If any symbol assignments refers to a skewed symbol then add that symbol to the set of skewed symbols.
+            * */
+            PlanNode child = node.getSource().accept(this, context);
+
+            // Symbols that will be removed by this projection.
+            Set<Symbol> outputSymbols = new HashSet<>(node.getOutputSymbols());
+            Set<Symbol> symbolsToRemove = new HashSet<>(node.getSource().getOutputSymbols());
+            symbolsToRemove.removeAll(outputSymbols);
+
+            // Add all symbols that refer to a skewed symbol.
+            node.getAssignments().entrySet()
+                    .forEach(entry -> {
+                        Expression value = entry.getValue();
+                        if (entry.getValue() instanceof SymbolReference) {
+                            Symbol skewedSymbol = Symbol.from(value);
+                            if (context.containsKey(Symbol.from(value)) && !context.containsKey(entry.getKey())) {
```

Comment:
This context cannot be global to the whole plan. It is technically possible for symbol names to get reused across unrelated plan nodes.

Reply:
Interesting!, i didn't know that. The context here is created independently for each child of a join node (left/right), and we don't allow nested joins. Does that still violate the assumption ?.